src/Shared/ChangeWaves.cs

```diff
@@ -0,0 +1,45 @@
+using Microsoft.Build.Shared;
+using System;
+
+namespace Microsoft.Build.Utilities
+{
+    public class ChangeWaves
+    {
+        public const string Wave16_8 = "16.8";
+        public const string Wave16_10 = "16.10";
+        public const string Wave17_0 = "17.0";
+
+        /// <summary>
+        /// Compares version against the MSBuildChangeWave environment variable.
+        /// Version MUST be of the format: "xx.yy".
+        /// </summary>
+        /// <param name="wave">The version to compare.</param>
+        /// <returns>A bool indicating whether the version is enabled.</returns>
+        public static bool IsChangeWaveEnabled(string wave)
+        {
+            // This is opt out behavior, all waves are enabled by default.
+            // If version is invalid, 
+            if (string.IsNullOrEmpty(Traits.Instance.MSBuildChangeWave))
+            {
+                return true;
+            }
+
+            Version currentDisabledWave;
+
+            if (!Version.TryParse(Traits.Instance.MSBuildChangeWave, out currentDisabledWave))
+            {
+                // how do I throw a warning here?
```

Comment:
Warning as in logger `BuildWarningEventArgs`? Throw a custom exception type, catch it in the nearest places that can log events, and convert it to a warning event there.

Reply:
I was hoping to avoid this, but it looks like that won't be possible. What was suggested in the PR review meeting was to return some enum `ChangeWavesReturnType.InvalidFormat`, then at each feature site we check what the return value was and log a warning based on that.

Reply:
That wouldn't work for features implemented in places which don't have access to a LoggingService.

An alternative to bubbling up a special exception would be to push down the LoggingService everywhere (potentially guised as a more generic logging interface, maybe). Then you can change the method to `public static bool IsChangeWaveEnabled(string wave, IWhateverLoggingInteface log)`

Reply:
I got around this by throwing warnings in `Evaluator.cs`, as I'm now setting `MSBUILDCHANGEWAVEVERSION` as a default property. The version gets sanitized there and a warning is thrown for out-of-rotation waves and invalid formats.